code/sound/ds.cpp

```diff
@@ -1444,6 +1477,169 @@ int ds_get_free_channel(float new_volume, int snd_id, int priority)
+
+			if (first_free_channel == -1) {
+				first_free_channel = i;
+			}
+			continue;
+		} else {
+			if ( chp->snd_id == snd_id ) {
+				instance_count++;
+
+				// looping or ambient soudns can't be preempted
+				if ((chp->looping == FALSE) && !chp->is_ambient)
+				{
+					if (chp->priority > least_important_instance_priority) {
+						least_important_instance_vol = chp->vol;
+						least_important_instance_priority = chp->priority;
+						least_important_instance_index = i;
+					} else if ((chp->priority == least_important_instance_priority)
+						 && (chp->vol < least_important_instance_vol)) {
+						least_important_instance_vol = chp->vol;
+						least_important_instance_priority = chp->priority;
+						least_important_instance_index = i;
+					}
+				}
+			} else if ( chp->is_voice_msg || (chp->looping == TRUE) || chp->is_ambient ) {
+				// a playing voice message, looping sound, or ambient sound is not allowed to be preempted.
+			} else if ( (chp->priority > least_important_priority) ) {
+				least_important_index = i;
+				least_important_vol = chp->vol;
+				least_important_priority = chp->priority;
+			} else if ( (chp->priority == least_important_priority) && (chp->vol < least_important_vol) ) {
+				least_important_index = i;
+				least_important_vol = chp->vol;
+				least_important_priority = chp->priority;
+			}
+		}
+	}
+
+	// If we've exceeded the limit, then stop the least important duplicate if it is lower or equal volume
+	// otherwise, don't play the sound
+	if ( ((unsigned int)instance_count >= enhanced_limit) && (least_important_instance_index >= 0) ) {
+		if (least_important_instance_vol <= new_volume) {
+			ds_close_channel_fast(least_important_instance_index);
+			first_free_channel =least_important_instance_index;
+		} else {
+			// don't exceed per-sound concurrency limits, even if there are spare channels
+			first_free_channel = -1;
+		}
+	} else if (first_free_channel == -1) {
+		// we haven't reached the limit, but we have run out of channels
+		// attempt to stop the least important sound to play our sound if priority demands it
+		if ( (least_important_index != -1)) {
+			// Check if the priority difference is great enough (must play or at least 2 levels away)
+			// and if the least important sound's volume playing is less than or equal to the volume of the requested sound.
+			// If so, then we are going to trash the least important sound.
+			if ( (enhanced_priority == SND_ENHANCED_PRIORITY_MUST_PLAY) || (Channels[least_important_index].priority - enhanced_priority >= 2)) {
+				if ( Channels[least_important_index].vol <= new_volume ) {
+					ds_close_channel_fast(least_important_index);
+					first_free_channel = least_important_index;
+				}
+			}
+		}
+	}
+
+	if ( (first_free_channel >= 0) && (Channels[first_free_channel].source_id == 0) ) {
+		OpenAL_ErrorCheck( alGenSources(1, &Channels[first_free_channel].source_id), return -1 );
+	}
+	return first_free_channel;
+}
+
+/**
```

Comment:
Apologies, I should have commented on the end, not start, of the comment block so these inline boxes were clearer.
I'm referring to the comment block that starts with "Generic function for getting a free channel".

It appears a pre-Doxygen layout, yet there is no function below it before the next Doxygen comment block which relates to ```ds_get_free_channel()```. If Doxygen-ified, can the older function description be removed as redundant?